src/ast/utility/Visitor.h

```diff
@@ -53,372 +53,116 @@
 #include "ast/UnnamedVariable.h"
 #include "ast/UserDefinedFunctor.h"
 #include "ast/Variable.h"
+#include "souffle/Visitor.h"
 #include "souffle/utility/FunctionalUtil.h"
 #include "souffle/utility/MiscUtil.h"
-#include <cstddef>
-#include <functional>
-#include <memory>
-#include <type_traits>
-#include <typeinfo>
-#include <vector>
 
 namespace souffle::ast {
-
-/** A tag type required for the is_ast_visitor type trait to identify AstVisitors */
-struct ast_visitor_tag {};
-
 /**
- * The generic base type of all AstVisitors realizing the dispatching of
- * visitor calls. Each visitor may define a return type R and a list of
- * extra parameters to be passed along with the visited Nodes to the
- * corresponding visitor function.
- *
- * @tparam R the result type produced by a visit call
- * @tparam Params extra parameters to be passed to the visit call
+ * The generic base type of all AstVisitors
+ * @see souffle::Visitor
  */
 template <typename R = void, typename NodeType = Node const, typename... Params>
-struct Visitor : public ast_visitor_tag {
-    /** A virtual destructor */
-    virtual ~Visitor() = default;
-
-    /** The main entry for the user allowing visitors to be utilized as functions */
-    R operator()(NodeType& node, Params const&... args) {
-        return visit(node, args...);
-    }
+struct Visitor : public souffle::Visitor<R, NodeType, Params...> {
+    using souffle::Visitor<R, NodeType, Params...>::visit_;
 
-    /**
-     * The main entry for a visit process conducting the dispatching of
-     * a visit to the various sub-types of Nodes. Sub-classes may override
-     * this implementation to conduct pre-visit operations.
-     *
-     * @param node the node to be visited
-     * @param args a list of extra parameters to be forwarded
-     */
-    virtual R visit(NodeType& node, Params const&... args) {
+    virtual R visit(NodeType& node, Params const&... args) override {
         // dispatch node processing based on dynamic type
 
-#define FORWARD(Kind) \
-    if (auto* n = as<Kind>(node)) return visit##Kind(*n, args...);
-
         // types
-        FORWARD(SubsetType);
-        FORWARD(UnionType);
-        FORWARD(RecordType);
-        FORWARD(AlgebraicDataType);
+        SOUFFLE_VISITOR_FORWARD(SubsetType);
```

Comment:
Do we really need such a long name?

Reply:
Since the macro is now "public' I was looking for a clear, unambiguous name.  However, happy to shorten if you have suggestions.